### DISCUSSION TITLE
Agents awareness of ACLK connection and liveness status from all its existing instances

### DISCUSSION THREAD

**Discussion by hugovalente-pm** [netdata/netdata#12120] [Ideas]
Hi @stelfrag @odynik 

For a new initiative we want to do on Agent, and to potentially support other features, there is the need for a given agent/node to have awareness about:
* what are the other node instances that exist of itself
* what is the live status of these in regards to Netdata Cloud in regards to:
  * claimed
  * liveness status

This will possibility mean propagating, through streaming, the connectivity to Netdata Cloud of every instance. We need to find way for each node instance on every agent (childs and parents) to know all the node instances it is being replicated (and currently connected) and their liveness in regards to Netdata Cloud connectivity.

For example, if there are 3 parents and 1 child, all 4 agents should know the presence of the other 3 and they know in real-time their connectivity to netdata.cloud.

What do you think of this?
  
cc @ktsaou

**Comment by ktsaou** [comment]
@stelfrag we have an alternative way to implement this, but if the agent has this knowledge the user experience would be a lot better. So, if this can happen at the agent we would prefer it.

**Reply by stelfrag** [reply]
yes it is better for the agents to communicate this via the streaming protocol. For now children will propagate the claim status (claim_id) to their parent. This needs to actually include the ACLK connection status (and changes to it). 

We need to add the information from the parent to the children (add that functionality to the protocol) or *quick* workaround for the child to poll the parent.

**Reply by ktsaou** [reply]
This needs to work for multi-parent setups too.
Do you believe that it is something simple to implement in the following 2 weeks? If not, we need to progress without it.

**Reply by stelfrag** [reply]
Multi-parent you mean something that can be configured now (not child to multiple parents concurrently) 

That is 

```mermaid
  graph TD;
      Child1-->Parent;
      Child2-->Proxy1;
      Proxy1-->Proxy2;
      Child3-->Proxy2;
      Proxy2-->Parent;
```

**Reply by ktsaou** [reply]
Yes, exactly.

Please don't make the children stream to multiple parents at the same time. The parents need to be a cluster and they need to sync with each other, with all children connecting to any one of them (but only one of them at a time). The existing protocol (with the planned enhancements to fill past gaps) should be able to do it quite easily by configuring "circular proxy configs" (i,e, two parents pointing to one another for their parent). I can explain more if needed.

**Reply by stelfrag** [reply]
We can probably have a way to have this information provided that a child node instance will not need to "act" in real time if its parent status changes. We can add a new command to the streaming protocol to propagate changes `upstream` but the other way around is a bit trickier (we are also working on it though) 

So an approach would be to have a status request on a child which in this case will issue a request to the parent (and its parent) to get the necessary information and will build a response based on that

This would, I guess,  be the *alternative* to try and do it directly by checking the api/v1/info and checking where that child is streaming to, provided there is access to the parent to run the query from the UI

What information will be needed in that response? Examples

- machine guid
- claim id
- node id
- hop count
- aclk connectivity

**Reply by hugovalente-pm** [reply]
The information you mention for the response @stelfrag seems 👌 

Regarding the approach, the time to get a response to the request that a child issues to the parent, will depend of the levels of hierarchy, if we have a node instance where hops is 4 the process would need to be propagated across 3 nodes until we get the final result to reach the child, but if we do it with `api/v1/info` requests it would be the same with (maybe) more potential breaking points.

So in end, between the approach where child requests to parent (and parent to its ancestor):
* a) the status request through ACLK against
* b) the `api/v1/info`
will depend on effort and potential risk on breaking points

**Reply by ktsaou** [reply]
I just learned there is also `/api/v1/aclk` that provides the same info as `netdatacli aclk-state`.

I have also opened the ticket #12217 to improve the info in both of them (@underhood told me both the API and the `netdatacli` output as the same function in the code.
